### ISSUE TITLE
Chebfun2 takes `O(n^2)` time to evaluate at `n` points

### ISSUE THREAD

**Issue by trefethen** [chebfun/chebfun#1487]
Suppose f is a fixed chebfun2.  You would think that the time to evaluate f(s,s), where s is a vector of length n, would be proportional to n, but @bhashemi and I have noticed that it seems to be proportional to n^2.  @ajt60gaibb, can you shed light on this?  If you try this at home, don't increase k to 2^15! -- on my desktop, at least, it hangs and I have to restart Matlab.
```
>> f = chebfun2(@(x,y) exp(x.*exp(y)));
>> for lk=10:14, k=2^lk; s=rand(k,1); tic, f(s,s); disp(k), toc, end
        1024
Elapsed time is 0.027070 seconds.
        2048
Elapsed time is 0.059683 seconds.
        4096
Elapsed time is 0.235066 seconds.
        8192
Elapsed time is 0.898304 seconds.
       16384
Elapsed time is 3.560397 seconds.
>> 
```

**Comment by ajt60gaibb** [comment]
Oh, I thought I checked for this situation... 

It is highly likely that the code is mistakenly evaluating the Chebfun2 at 
the `n^2` points  `(s(j),s(k))` and then throw away all of them apart from 
those on the diagonal.  

This can be easily fixed by checking the size of the input arguments to 
feval.

**Comment by bhashemi** [comment]
This is indeed the case. Here is an easy fix by applying the following changes to Lines 85-89 of `chebfun2/feval`:

``
    else 
%        takeDiag = 1; % B.H.
        takeDiag = 0; % B.H.
    end
    
    % Evaluate:
    if ( min(size(x)) == 1 && all(size(x) == size(y)) ) % B.H.
        out = feval(cols, y(:)) .* feval(rows, x(:)) * diag(D); % B.H.
    else % B.H.
        out = feval(cols, y(:)) * D * feval(rows, x(:)).';
    end % B.H.
``
But, a more fundamental question is: In the light of this issue, do we need the "takeDiag" flag at all?

**Comment by ajt60gaibb** [comment]
This has been fixed in #1502. 

```
f = chebfun2(@(x,y) exp(x.*exp(y)));
for lk=10:14, k=2^lk; s=rand(k,1); tic, f(s,s); disp(k), toc, end
        1024

Elapsed time is 0.019190 seconds.
        2048

Elapsed time is 0.035747 seconds.
        4096

Elapsed time is 0.055593 seconds.
        8192

Elapsed time is 0.188746 seconds.
       16384

Elapsed time is 0.271945 seconds.

```
